### PR TITLE
Async amax reduction

### PR DESCRIPTION
**Pull request by erhoo82**
- Support async AMAX reduction
- AMAX reduction is default enabled and can be disabled by setting an env var (`NVTE_ASYNC_AMAX_REDUCTION`)

### CODE REVIEW

**Comment by ptrendx** [comment]
@erhoo82 Please sign your commits (please refer to CONTRIBUTING.rst to see how to do this).

**Comment by erhoo82** [comment]
Thanks, signed the commit.

File: transformer_engine/pytorch/module.py [hunk 1]
```diff
@@ -105,6 +109,10 @@ def get_workspace() -> torch.Tensor:
 def _prepare_backward(fp8: bool, fp8_meta: Dict[str, Any],  name: str = "") -> None:
     """Checks and prep for BWD."""
     if fp8:
+        global _amax_reduce_handle_bwd
+        if is_last_fp8_module() and _amax_reduce_handle_bwd is not None:
+            _amax_reduce_handle_bwd.wait()
```
Comment (timmoon10) [inline transformer_engine/pytorch/module.py:114]: I'm not sure I fully understand this PR, but the goal is to launch an async amax reduction when doing the backward pass for the first FP8 module and synchronize before the next backward pass (i.e. the backward pass for the last FP8 module)? Could we accomplish the same behavior by resetting the request object after synchronizing?
```suggestion
        global _amax_reduce_handle_bwd
        if  _amax_reduce_handle_bwd is not None:
            _amax_reduce_handle_bwd.wait()
            _amax_reduce_handle_bwd = None
```
This would be somewhat simpler since it doesn't need the `is_last_fp8_module` infrastructure.

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:114]: I believe the last module is to keep consistent with is_first_module. I'll test this feature once.

Comment (erhoo82) [reply transformer_engine/pytorch/module.py:114]: The goal of this PR is (1) overlap amax reudction with p2p in both pipeline warmup and flush phases, (2) overlap amax reduction with backprop (in case of the reduction after fwd microbatch) and fprop (in case of the reduction after bwd microbatch).

Doesn't your suggested change actually do sync after each Transformer backward not after the last one?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:114]: This change is not a standalone change and `_amax_reduce_handle_bwd` would need to be set properly each time IIUC.

File: transformer_engine/pytorch/fp8.py [hunk 2]
```diff
@@ -496,16 +520,18 @@ def get_fp8_te_dtype(
 
 
 def reduce_tensor_across_group_op_max(
-    tensor: torch.Tensor, group: dist_group_type
+    tensor: torch.Tensor, group: dist_group_type, async_amax_reduction: bool
```
Comment (ksivaman) [inline transformer_engine/pytorch/fp8.py:523]:
```suggestion
    tensor: torch.Tensor, group: dist_group_type, async_op: bool
```

File: transformer_engine/pytorch/fp8.py [hunk 3]
```diff
@@ -496,16 +520,18 @@ def get_fp8_te_dtype(
...
 ) -> None:
     """Reduce tensor across given group."""
     if torch.distributed.is_initialized():
-        torch.distributed.all_reduce(
+        wait_handle = torch.distributed.all_reduce(
             tensor,
             op=torch.distributed.ReduceOp.MAX,
             group=group,
-            async_op=False,
+            async_op=async_amax_reduction,
```
Comment (ksivaman) [inline transformer_engine/pytorch/fp8.py:531]:
```suggestion
            async_op=async_op,
```

File: transformer_engine/pytorch/fp8.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/fp8.py:523]: This is a general function for reducing a tensor across the given group so the argument names should not be specific to amax reduction.

**Comment by erhoo82** [comment]
@ksivaman 
I tested and confirmed the functionality.

File: transformer_engine/pytorch/fp8.py (hunk 2 above)
Comment (erhoo82) [reply transformer_engine/pytorch/fp8.py]: Updated

File: transformer_engine/pytorch/fp8.py (hunk 3 above)
Comment (erhoo82) [reply transformer_engine/pytorch/fp8.py]: Done

File: transformer_engine/pytorch/fp8.py [hunk 4]
```diff
@@ -536,7 +536,7 @@ def global_amax_reduction(
     wait_handle = reduce_tensor_across_group_op_max(
         contiguous_amax,
         fp8_meta["fp8_group"],
-        fp8_meta["async_amax_reduction"],
+        fp8_meta["async_op"],
```
Comment (ksivaman) [inline transformer_engine/pytorch/fp8.py:539]: I'm nitpicking, but this variable was better off before, i.e. "async_amax_reduction". My name change suggestion was specifically for the `reduce_tensor_across_group_op_max` function because it that function not specifically related to amax reduction.

File: transformer_engine/pytorch/module.py [hunk 5]
```diff
@@ -191,7 +191,7 @@ def __init__(self) -> None:
         self.sequence_parallel = False
         self.fp8_weight_shapes = []
         self.fp8_meta["autocast_id_fwd_stack"] = []
-        self.fp8_meta["async_amax_reduction"] = bool(
+        self.fp8_meta["async_op"] = bool(
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:194]: Same as above; `async_op` &rarr; `async_amax_reduction`.

File: transformer_engine/pytorch/fp8.py [hunk 6]
```diff
@@ -518,14 +528,19 @@ def global_amax_reduction(
...
+        return None
 
     chunk_sizes = [x.numel() for x in _global_fp8_buffer[amax_buffer_key]]
     contiguous_amax = torch.cat(_global_fp8_buffer[amax_buffer_key])
 
-    reduce_tensor_across_group_op_max(contiguous_amax, fp8_meta["fp8_group"])
+    wait_handle = reduce_tensor_across_group_op_max(
+        contiguous_amax,
+        fp8_meta["fp8_group"],
+        fp8_meta["async_op"],
```
Comment (ksivaman) [inline transformer_engine/pytorch/fp8.py:539]:
```suggestion
        fp8_meta["async_amax_reduction"],
```

File: transformer_engine/pytorch/module.py [hunk 7]
```diff
@@ -184,6 +191,9 @@ def __init__(self) -> None:
         self.sequence_parallel = False
         self.fp8_weight_shapes = []
         self.fp8_meta["autocast_id_fwd_stack"] = []
+        self.fp8_meta["async_op"] = bool(
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:194]:
```suggestion
        self.fp8_meta["async_amax_reduction"] = bool(
```

**Review by ksivaman — APPROVED** [review]
LGTM, thanks!

**Comment by Victarry** [comment]
Hello, @ksivaman @erhoo82 , I think it's more reasonable to put `amax_reduce_handle_fwd.wait() ` before `FP8GlobalStateManager.copy_amax_from_global_buffer(self.fp8_meta, forward=True)`, which ensures the reduction is finished before copy amax from global buffer.


https://github.com/NVIDIA/TransformerEngine/blob/b8ba734e01671e77d7efebc1f773adf12839de9d/transformer_engine/pytorch/module/base.py#L582-L616
